File: 1000 and 1 tasks/data-structure1/valid-sudoku.cs
public class Solution {

    public bool IsCharValid(char x) {
        return x >= '1' && x <= '9' || x == '.';
    }

    public bool IsValidSudoku(char[][] board) {
        var columns = new bool[9][];
        var grids = new bool[9][];

        for (int i = 0; i < 9; i++) {
            columns[i] = new bool[10];
            grids[i] = new bool[10];
        }

        for (int i = 0; i < 9; i++) {
            var line = new bool[10];
            for (int j = 0; j < 9; j++) {
                var chr = board[i][j];
                if (!IsCharValid(chr)) {
                    return false;
                }
                if (char.IsDigit(chr)) {
                    if (line[chr - '0']) {
                        return false;
                    }
                    line[chr - '0'] = true;
                    if (columns[j][chr-'0']) {
                        return false;
                    }
                    if (grids[i / 3 * 3 + j / 3][chr-'0']) {
                        return false;
                    }
                    grids[i / 3 * 3 + j / 3][chr-'0'] = true;
                    columns[j][chr-'0'] = true;
                }
            }

        }

        return true;
    }
}

File: 1000 and 1 tasks/new-top-150/valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {

        var nrows = board.length;
        var ncols = board[0].length;

        for (int i = 0; i < nrows; i++) {
            var rowsState = new boolean[10];
            for (var j = 0; j < ncols; j++) {
                if (board[i][j] == '.') {
                    continue;
                }

                var x = board[i][j] - '0';
                if (rowsState[x]) {
                    return false;
                }
                rowsState[x] = true;
            }
        }

        for (int j = 0; j < ncols; j++) {
            var colState = new boolean[10];
            for (var i = 0; i < nrows; i++) {
                if (board[i][j] == '.') {
                    continue;
                }

                var x = board[i][j] - '0';
                if (colState[x]) {
                    return false;
                }
                colState[x] = true;
            }
        }

        var squareStates = new boolean[10][10];

        for (int i = 0; i < nrows; i++) {
            for (var j = 0; j < ncols; j++) {
                if (board[i][j] == '.') {
                    continue;
                }
                // line *3 + col
                var line = i / 3;
                var col = j / 3;
                var squareIndex = line * 3 + col;

                var x = board[i][j] - '0';
                if (squareStates[squareIndex][x]) {
                    return false;
                }
                squareStates[squareIndex][x] = true;
            }
        }

        // todo check squares

        return true;
    }
}